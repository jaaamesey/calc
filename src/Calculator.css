#result {
    height: 30px;
}

/*# sourceMappingURL=Calculator.css.map */
