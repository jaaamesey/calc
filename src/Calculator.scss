#result {
  height: 30px;
}
