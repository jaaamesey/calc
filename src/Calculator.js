import React from 'react';
import 'mathjs';
import './App.scss';
import "./Calculator.scss";
import {Button, ButtonGroup, Dropdown, FormControl} from "react-bootstrap"
import {all, create} from "mathjs";
import insertText from 'insert-text-textarea';

export class Calculator extends React.Component {
    constructor() {
        super();
        this.mathJsConfig = {
            epsilon: 1e-12,
            matrix: 'Matrix',
            number: 'BigNumber',
            precision: 128,
            predictable: false,
            randomSeed: null
        };
        this.math = create(all, this.mathJsConfig);

        this.state = {
            expression: ' ',
            result: ''
        };
        this.textInput = React.createRef();
    }

    renderCalculatorButtons = () =>
        (
            <div>
                <FormControl ref={this.textInput} value={this.state.expression} onChange={this.onExpressionChanged}
                             placeholder="" aria-label="Expression"/>
                <div id="result">{this.state.result}</div>
                <ButtonGroup>
                    <Button onClick={this.moveCursorLeft} variant="secondary">
                        {'<'}
                    </Button>
                    <Button onClick={this.append.bind(this, '-')} variant="secondary">
                        AC
                    </Button>
                    <Button onClick={this.clear} variant="secondary">
                        C
                    </Button>
                    <Button onClick={this.moveCursorRight} variant="secondary">
                        {'>'}
                    </Button>
                </ButtonGroup>
                <br/>
                <ButtonGroup>
                    <Button onClick={this.append.bind(this, '+')} variant="secondary">
                        +
                    </Button>
                    <Button onClick={this.append.bind(this, '-')} variant="secondary">
                        -
                    </Button>
                    <Button onClick={this.append.bind(this, '*')} variant="secondary">
                        x
                    </Button>
                    <Button onClick={this.append.bind(this, '/')} variant="secondary">
                        รท
                    </Button>
                </ButtonGroup>
                <Dropdown>
                    <Dropdown.Toggle variant="success" id="dropdown-basic">
                        Dropdown Button
                    </Dropdown.Toggle>

                    <Dropdown.Menu>
                        <Dropdown.Item href="test">Action</Dropdown.Item>
                        <Dropdown.Item href="#/action-2">Another action</Dropdown.Item>
                        <Dropdown.Item href="#/action-3">Something else</Dropdown.Item>
                    </Dropdown.Menu>
                </Dropdown>
            </div>
        );

    render = () => {
        return this.renderCalculatorButtons();
    };

    append = (str) => {
        this.textInput.current.focus();
        insertText(this.textInput.current, str);
    };

    onExpressionChanged = (event) => {
        let expression = event.target.value.replace(/\s+/g, ' ');
        this.setState({
            expression: expression,
            result: this.tryEnumerate(expression)
        });
    };

    clear = () => {
        this.textInput.current.focus();
        this.textInput.current.selectionStart = 0;
        this.textInput.current.selectionEnd = this.state.expression.length;
        insertText(this.textInput.current, '');
        this.textInput.current.focus();
    };

    moveCursorLeft = () => {
        this.textInput.current.focus();
        if (this.textInput.current.selectionStart === this.textInput.current.selectionEnd)
            this.textInput.current.selectionStart -= 1;
        this.textInput.current.selectionEnd = this.textInput.current.selectionStart;
    };
    moveCursorRight = () => {
        this.textInput.current.focus();
        if (this.textInput.current.selectionStart === this.textInput.current.selectionEnd)
            this.textInput.current.selectionEnd += 1;
        this.textInput.current.selectionStart = this.textInput.current.selectionEnd;
    };

    tryEnumerate(exp) {
        exp = exp.trim();
        if (exp === "")
            return "";
        try {
            const result = this.math.compile(exp).evaluate().toString();
            // Handle strange bug with function names
            if (result.includes("function fn(arg")) {
                return "ERROR";
            }
            return result;
        } catch {
            return "ERROR";
        }
    }

}
