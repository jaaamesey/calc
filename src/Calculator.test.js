import React from 'react';
import ReactDOM from "react-dom";

import {Calculator} from "./Calculator";


it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<Calculator/>, div);
});

it('adds numbers correctly', () => {
    const component = new Calculator();
    expect(component.tryEnumerate('2 + 6')).toBe('8');
});
