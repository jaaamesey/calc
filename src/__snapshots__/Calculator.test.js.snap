// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`adds numbers correctly 1`] = `
<div>
  <input
    aria-label="Expression"
    className="form-control"
    onChange={[Function]}
    placeholder=""
    value=" "
  />
  <div
    id="result"
  >
    
  </div>
  <div
    className="btn-group"
    role="group"
  >
    <button
      className="btn btn-secondary"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      &lt;
    </button>
    <button
      className="btn btn-secondary"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      AC
    </button>
    <button
      className="btn btn-secondary"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      C
    </button>
    <button
      className="btn btn-secondary"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      &gt;
    </button>
  </div>
  <br />
  <div
    className="btn-group"
    role="group"
  >
    <button
      className="btn btn-secondary"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      +
    </button>
    <button
      className="btn btn-secondary"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      -
    </button>
    <button
      className="btn btn-secondary"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      x
    </button>
    <button
      className="btn btn-secondary"
      disabled={false}
      onClick={[Function]}
      type="button"
    >
      รท
    </button>
  </div>
  <div
    className="dropdown"
    onKeyDown={[Function]}
  >
    <button
      aria-expanded={false}
      aria-haspopup={true}
      className="dropdown-toggle btn btn-success"
      disabled={false}
      id="dropdown-basic"
      onClick={[Function]}
      type="button"
    >
      Dropdown Button
    </button>
  </div>
</div>
`;
